### PR TITLE
SE: Reset static fields on static method calls

### LINKED ISSUE #6100: S2259: Reset constraints on calls in static methods
Same as calls with a `this` receiver (aka instance methods) reset field constraints, calls to static methods should also reset field constraints:
```cs
public class C
{
  static object field;
  static void M()
  {
    field = null; // learn about field
    OtherStaticMethod();
    field.ToString(); // Compliant
  }
}
```

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #6100

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SymbolicExecution/SETestContext.cs [hunk 1]
```diff
@@ -137,10 +139,6 @@ public void Main(bool boolParameter{additionalParameters})
     public NotImplementedException CreateException() => new NotImplementedException();
     public void InstanceMethod(object parameter = null) {{ }}
     public static void StaticMethod() {{ }}
-
-    private void Tag(string name) {{ }}
-    private static void Tag(string name, object arg = null) {{ }}
-    private static T Unknown<T>() => default;
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SymbolicExecution/SETestContext.cs:143]: I moved the methods further away from the Sample class. I hope the invocations to Tag no longer interfere with our logic now and in the future.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 2]
```diff
@@ -216,6 +215,28 @@ public void Invocation_InstanceMethodCallDoesNotClearFieldForOtherAccess(string
             validator.ValidateTag("AfterStaticField", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
         }
 
+        [DataTestMethod]
+        [DataRow("StaticMethod();")]
+        [DataRow("Sample.StaticMethod();")]
+        public void Invocation_StaticMethodCallDoesNotClearFieldForOtherAccess(string invocation)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:221]: I don't clearly understand the name. Does it mean this?
```suggestion
        public void Invocation_StaticMethodCall_DoesNotClearInstanceFields(string invocation)
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 3]
```diff
@@ -341,6 +362,95 @@ public void Instance_InstanceMethodCallClearsField()
                 new SymbolicValue().WithConstraint(ObjectConstraint.NotNull));
         }
 
+        [TestMethod]
+        public void Invocation_StaticMethodCallClearsField()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:366]: Rename & let's move it up, near the other `Invocation_StaticMethodCall_Xxxx`
```suggestion
        public void Invocation_StaticMethodCall_ClearsField()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 4]
```diff
@@ -341,6 +362,95 @@ public void Instance_InstanceMethodCallClearsField()
...
+{
+    protected static object _A;
+    public static void BaseMethod() { }
+}
+public class Other
+{
+    public static object _B;
+    public static void OtherMethod() { }
+}
+";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:428]: Nitpick
```suggestion
}";
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 5]
```diff
@@ -341,6 +362,95 @@ public void Instance_InstanceMethodCallClearsField()
                 new SymbolicValue().WithConstraint(ObjectConstraint.NotNull));
         }
 
+        [TestMethod]
+        public void Invocation_StaticMethodCallClearsField()
+        {
+            var code = @"
+public class Sample: Base
+{
+    public static object _C1;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:371]: We don't need the underscores. I'd prefer more descriptive names like 'FieldThis1`, `FieldThis2`, `FieldBase`, `FieldOtherClass`

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:371]: .. or `SampleField1`, to be aligned with `Sample` class and `SampleMethod`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 6]
```diff
@@ -341,6 +362,95 @@ public void Instance_InstanceMethodCallClearsField()
...
+            validator.ValidateTag("SampleMethod_C1", x => x.HasConstraint(ObjectConstraint.Null).Should().BeFalse());
+            validator.ValidateTag("SampleMethod_C2", x => x.HasConstraint(ObjectConstraint.Null).Should().BeFalse());
+
+            // OtherMethod() resets only its own constraints
+            validator.ValidateTag("OtherMethod_Base_A", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("OtherMethod_Other_B", x => x.HasConstraint(ObjectConstraint.Null).Should().BeFalse());
+            validator.ValidateTag("OtherMethod_C1", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("OtherMethod_C2", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+
+            // BaseMethod() called from Sample only resets Base field
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:447]: As long as SOLID is respected 😆

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor polishing

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs]: I moved it after the "Instance" tests and renamed the "Instance" tests also.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs [hunk 7]
```diff
@@ -51,6 +51,10 @@ protected override ProgramState[] Process(SymbolicContext context, IInvocationOp
         {
             state = state.ResetFieldConstraints();
         }
+        if (invocation.TargetMethod.IsStatic)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:54]: I'm not sure, how we should handle extension methods. We already do check that the `ContainingType is the same, so only calls from one extension method to another extension method of the same class will be ever processed and there we would reset static fields stored in the extension method class (whoever does such things?).
However, the current behavior is not documented by tests by now.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:54]: Extension method is a very specific case that should not touch the current class, so we're fine. I'd leave it as it is.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**